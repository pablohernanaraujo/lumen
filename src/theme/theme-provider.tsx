import React, {
  createContext,
  type FC,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';
import { useColorScheme } from 'react-native';

import { colorTokens } from './colors';
import { baseTokens } from './tokens';
import type {
  Mode,
  Theme,
  ThemeContextValue,
  ThemeProviderProps,
} from './types';

const ThemeContext = createContext<ThemeContextValue | null>(null);

export const ThemeProvider: FC<ThemeProviderProps> = ({
  children,
  preference,
  persistenceKey = '@lumen-theme',
}) => {
  const systemMode = useColorScheme() as Mode | null;
  const [userMode, setUserMode] = useState<Mode | null>(preference || null);

  const currentMode: Mode = userMode ?? systemMode ?? 'light';

  const theme = useMemo<Theme>(
    () => ({
      ...baseTokens,
      colors: colorTokens[currentMode],
      mode: currentMode,
    }),
    [currentMode],
  );

  const setMode = useCallback((mode: Mode): void => {
    setUserMode(mode);
  }, []);

  const toggleMode = useCallback((): void => {
    setUserMode(currentMode === 'light' ? 'dark' : 'light');
  }, [currentMode]);

  const contextValue = useMemo<ThemeContextValue>(
    () => ({
      theme,
      mode: currentMode,
      setMode,
      toggleMode,
    }),
    [theme, currentMode, setMode, toggleMode],
  );

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeContextValue => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
