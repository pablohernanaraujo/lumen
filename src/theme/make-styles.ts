import { StyleSheet } from 'react-native';

import { useTheme } from './theme-provider';
import type { NamedStyles, StylesOrFunction, Theme } from './types';

export const makeStyles =
  <T extends NamedStyles<T>>(
    factory: StylesOrFunction<T>,
  ): (() => StyleSheet.NamedStyles<T>) =>
  (): StyleSheet.NamedStyles<T> => {
    const { theme } = useTheme();

    const styles =
      typeof factory === 'function'
        ? (factory as (theme: Theme) => T)(theme)
        : factory;

    return StyleSheet.create(styles);
  };
