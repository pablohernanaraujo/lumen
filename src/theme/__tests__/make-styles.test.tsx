/* eslint-disable max-nested-callbacks */
import React from 'react';
import { Text, View } from 'react-native';
import { render } from '@testing-library/react-native';

import { makeStyles } from '../make-styles';
import { ThemeProvider } from '../theme-provider';

const TestComponent: React.FC = () => {
  const styles = makeStyles((theme) => ({
    container: {
      backgroundColor: theme.colors.background,
      padding: theme.spacing.md,
    },
    text: {
      color: theme.colors.text.primary,
      fontSize: theme.typography.size.md,
    },
    dynamic: {
      borderWidth: 1,
      borderColor: theme.colors.border,
    },
  }))();

  return (
    <View style={styles.container}>
      <Text style={styles.text}>Test Text</Text>
      <View style={styles.dynamic} />
    </View>
  );
};

describe('makeStyles', () => {
  describe('theme integration', () => {
    it('should create styles with light theme', () => {
      const { toJSON } = render(
        <ThemeProvider preference="light">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });

    it('should create styles with dark theme', () => {
      const { toJSON } = render(
        <ThemeProvider preference="dark">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('style function behavior', () => {
    it('should create consistent styles across multiple calls', () => {
      const styleFunction = makeStyles((theme) => ({
        test: {
          color: theme.colors.text.primary,
        },
      }));

      const TestConsistencyComponent: React.FC = () => {
        const styles1 = styleFunction();
        const styles2 = styleFunction();

        return (
          <View>
            <Text style={styles1.test}>Test 1</Text>
            <Text style={styles2.test}>Test 2</Text>
          </View>
        );
      };

      const { toJSON } = render(
        <ThemeProvider preference="light">
          <TestConsistencyComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });

    it('should handle complex style objects', () => {
      const ComplexStyleComponent: React.FC = () => {
        const styles = makeStyles((theme) => ({
          complex: {
            backgroundColor: theme.colors.background,
            borderRadius: theme.radii.md,
            paddingHorizontal: theme.spacing.lg,
            paddingVertical: theme.spacing.sm,
          },
        }))();

        return <View style={styles.complex} />;
      };

      const { toJSON } = render(
        <ThemeProvider preference="light">
          <ComplexStyleComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });

    it('should handle empty style objects', () => {
      const EmptyStyleComponent: React.FC = () => {
        const styles = makeStyles(() => ({}))();

        return <View style={styles} />;
      };

      const { toJSON } = render(
        <ThemeProvider preference="light">
          <EmptyStyleComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('theme responsiveness', () => {
    it('should update styles when theme changes', () => {
      const ResponsiveComponent: React.FC<{ mode: 'light' | 'dark' }> = ({
        mode,
      }) => {
        const styles = makeStyles((theme) => ({
          responsive: {
            backgroundColor: theme.colors.background,
            color: theme.colors.text.primary,
          },
        }))();

        return (
          <Text style={styles.responsive} testID="responsive-text">
            Responsive Text
          </Text>
        );
      };

      const { rerender, toJSON: toJSONLight } = render(
        <ThemeProvider preference="light">
          <ResponsiveComponent mode="light" />
        </ThemeProvider>,
      );

      expect(toJSONLight()).toMatchSnapshot('light-responsive');

      rerender(
        <ThemeProvider preference="dark">
          <ResponsiveComponent mode="dark" />
        </ThemeProvider>,
      );

      expect(toJSONLight()).toMatchSnapshot('dark-responsive');
    });
  });
});
