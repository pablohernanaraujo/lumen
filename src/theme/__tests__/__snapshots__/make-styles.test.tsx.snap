// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeStyles style function behavior should create consistent styles across multiple calls 1`] = `
<View>
  <Text
    style={
      {
        "color": "#283238",
      }
    }
  >
    Test 1
  </Text>
  <Text
    style={
      {
        "color": "#283238",
      }
    }
  >
    Test 2
  </Text>
</View>
`;

exports[`makeStyles style function behavior should handle complex style objects 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 8,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
/>
`;

exports[`makeStyles style function behavior should handle empty style objects 1`] = `
<View
  style={{}}
/>
`;

exports[`makeStyles theme integration should create styles with dark theme 1`] = `
<View
  style={
    {
      "backgroundColor": "#000000",
      "padding": 16,
    }
  }
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontSize": 16,
      }
    }
  >
    Test Text
  </Text>
  <View
    style={
      {
        "borderColor": "#374151",
        "borderWidth": 1,
      }
    }
  />
</View>
`;

exports[`makeStyles theme integration should create styles with light theme 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "padding": 16,
    }
  }
>
  <Text
    style={
      {
        "color": "#283238",
        "fontSize": 16,
      }
    }
  >
    Test Text
  </Text>
  <View
    style={
      {
        "borderColor": "#E5E7EB",
        "borderWidth": 1,
      }
    }
  />
</View>
`;

exports[`makeStyles theme responsiveness should update styles when theme changes: dark-responsive 1`] = `
<Text
  style={
    {
      "backgroundColor": "#FFFFFF",
      "color": "#283238",
    }
  }
  testID="responsive-text"
>
  Responsive Text
</Text>
`;

exports[`makeStyles theme responsiveness should update styles when theme changes: light-responsive 1`] = `
<Text
  style={
    {
      "backgroundColor": "#FFFFFF",
      "color": "#283238",
    }
  }
  testID="responsive-text"
>
  Responsive Text
</Text>
`;
